SwiftCompilerSources/Sources/Optimizer/Utilities/WalkUtils.swift

```diff
@@ -0,0 +1,376 @@
+//===--- WalkUtils.swift - Utilities for use-def def-use walks ------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+func TODO(_ message: String = "Unhandled TODO", fun: String = #function,
+          file: String = #file, line: Int = #line) -> Never {
+  fatalError("\(#function)\(file):\(line) \(fun) \(message)")
+}
+
+func FATAL(_ message: String = "", fun: String = #function,
+          file: String = #file, line: Int = #line) -> Never {
+  fatalError("\(#function)\(file):\(line) \(fun) \(message)")
+}
+
+public enum WalkResult {
+  /// skip traversal
+  case ignore
+  /// stop the walk, add the current use/def and path to the result
+  case abortWalk
+  /// continue the walk
+  case continueWalk
+}
+
+protocol MemoizedUp {
+  mutating func shouldRecomputeUp(value def: Value, path: SmallProjectionPath) -> SmallProjectionPath?
+}
+
+protocol MemoizedDown {
+  mutating func shouldRecomputeDown(value def: Value, path: SmallProjectionPath) -> SmallProjectionPath?
+}
+
+protocol ValueDefUseWalker : MemoizedDown {
```

Comment:
What do you think about just having 2 protocols: `DefUseWalker` and `UseDefWalker`, both containing two walk-functions for values and addresses? I find this simpler to understand than encoding both dimensions (up-down, value-address) in the protocol name.
Then you also don't need to factor out the Memoize protocols.